File: openlmt/python/src/openlmt/otel/__tracing.py
# pylint: disable=line-too-long
"""
Setups up OpenTelemetry tracer
"""

import os
from opentelemetry import trace
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, SimpleSpanProcessor, ConsoleSpanExporter
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

# Global flag to check if the tracer provider initialization is complete.
TRACER_SET = False

def setup_tracing(application_name="default", tracer=None, otlp_endpoint=None, otlp_headers=None, disable_batch=False):
    """
    Sets up tracing with OpenTelemetry. Initializes the tracer provider and configures the span processor and exporter.

    Params:
        application_name (str): The name of the application to be used in traces.
        tracer (Tracer): Optional custom tracer. If provided, it is immediately returned and no setup is performed.
        otlp_endpoint (str): The OTLP exporter endpoint. Falls back to the OTEL_EXPORTER_OTLP_ENDPOINT environment variable if not specified.
        otlp_headers (dict): Headers for the OTLP request. Falls back to the OTEL_EXPORTER_OTLP_HEADERS environment variable if not specified.
        disable_batch (bool): Flag to disable the batch span processor in favor of a simpler processor for exporting.

    Returns:
        The provided custom tracer if not None; otherwise, a tracer instance configured according to the given parameters or environment variables.
    """

    # If an external tracer is provided, return it immediately.
    if tracer is not None:
        return tracer

    # Proceed with setting up a new tracer or configuration only if TRACER_SET is False.
    # pylint: disable=global-statement
    global TRACER_SET

    try:
        if not TRACER_SET:
            # Create a resource with the service name attribute.
            resource = Resource(attributes={SERVICE_NAME: application_name})

            # Initialize the TracerProvider with the created resource.
            trace.set_tracer_provider(TracerProvider(resource=resource))

            otlp_endpoint = otlp_endpoint or os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
            otlp_headers = otlp_headers or os.getenv("OTEL_EXPORTER_OTLP_HEADERS")

            # Configure span exporter and processor based on provided parameters or defaults.
            if otlp_endpoint:
                span_exporter = OTLPSpanExporter(endpoint=otlp_endpoint, headers=otlp_headers)
                span_processor = BatchSpanProcessor(span_exporter) if not disable_batch else SimpleSpanProcessor(span_exporter)
            else:
                span_exporter = ConsoleSpanExporter()
                span_processor = SimpleSpanProcessor(span_exporter)

            trace.get_tracer_provider().add_span_processor(span_processor)

            TRACER_SET = True

        return trace.get_tracer(__name__)

    # pylint: disable=bare-except
    except:
        return None

File: sdk/typescript/src/otel/tracing.ts
import { trace } from '@opentelemetry/api';
import { BasicTracerProvider } from '@opentelemetry/sdk-trace-base';
import { SetupTracerOptions } from '../types';
import {
  NodeTracerProvider,
  BatchSpanProcessor,
  ConsoleSpanExporter,
  SimpleSpanProcessor,
  SpanExporter,
} from '@opentelemetry/sdk-trace-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';

import Instrumentations from '../instrumentation';
import OpenlitConfig from '../config';
import { getRegisteredTracerProvider, parseExporters } from './utils';

let TRACER_SET = false;

export default class Tracing {
  static traceProvider: BasicTracerProvider;
  static traceExporter: OTLPTraceExporter | undefined;

  static setup(options: SetupTracerOptions) {
    // If an external tracer is provided, return it immediately (Python parity).
    if (options.tracer) return options.tracer;

    try {
      if (!TRACER_SET) {
        const existingProvider = getRegisteredTracerProvider();

        if (existingProvider) {
          // Reuse the host app's SDK TracerProvider — do not register a second one.
          this.traceProvider = existingProvider;
        } else {
          // No SDK provider configured yet — create one.
          const spanProcessors = Tracing.buildSpanProcessors(options);

          const provider = new NodeTracerProvider({
            resource: options.resource,
            spanProcessors,
          });

          provider.register();
          this.traceProvider = provider;
        }

        OpenlitConfig.updateConfig({
          ...options,
          tracer: (options.tracer || Tracing.traceProvider) as NodeTracerProvider,
        });

        Instrumentations.setup(
          Tracing.traceProvider,
          options?.disabledInstrumentors,
          options?.instrumentations
        );

        TRACER_SET = true;
      }

      return trace.getTracer('openlit');
    } catch (e) {
      console.error('[Traces] Failed to initialize traces:', e);
      return null;
    }
  }

  private static buildSpanProcessors(options: SetupTracerOptions) {
    const processors: (SimpleSpanProcessor | BatchSpanProcessor)[] = [];
    const exporterList = parseExporters('OTEL_TRACES_EXPORTER');

    if (exporterList) {
      for (const name of exporterList) {
        if (name === 'otlp') {
          const exporter = Tracing.createOTLPExporter(options);
          this.traceExporter = exporter;
          processors.push(Tracing.wrapProcessor(exporter, options.disableBatch));
        } else if (name === 'console') {
          processors.push(new SimpleSpanProcessor(new ConsoleSpanExporter()));
        }
      }
    } else {
      // Default: OTLP if endpoint is set, otherwise Console
      if (options.otlpEndpoint) {
        const exporter = Tracing.createOTLPExporter(options);
        this.traceExporter = exporter;
        processors.push(Tracing.wrapProcessor(exporter, options.disableBatch));
      } else {
        processors.push(new SimpleSpanProcessor(new ConsoleSpanExporter()));
      }
    }

    return processors;
  }

  private static createOTLPExporter(options: SetupTracerOptions): OTLPTraceExporter {
    const url = (options.otlpEndpoint || '') + '/v1/traces';
    return new OTLPTraceExporter({
      url,
      headers: options.otlpHeaders as Record<string, string> | undefined,
    });
  }

  private static wrapProcessor(
    exporter: SpanExporter,
    disableBatch: boolean
  ): SimpleSpanProcessor | BatchSpanProcessor {
    return disableBatch
      ? new SimpleSpanProcessor(exporter)
      : new BatchSpanProcessor(exporter);
  }

  static resetForTesting() {
    TRACER_SET = false;
  }
}
